### PR TITLE
executor: don't fill generated columns at the non-public state

### LINKED ISSUE #31735: connection running loop panic when add generated column for partition table
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1、run add generated column for partition table

### 2. What did you expect to see? (Required)
connection is normal

### 3. What did you see instead (Required)
connection running loop panic
Insert on duplicate failed, Error 1105: runtime error: index out of range [0] with length 0
[mysql] 2022/01/17 07:40:12 packets.go:123: closing bad idle connection: EOF

[2022/01/17 07:32:32.890 +00:00] [ERROR] [conn.go:1031] ["connection running loop panic"] [conn=96495] [lastSQL="insert into test_oncall2929.qhad_46 (create_date, atime, aid, m_req_all, text) values ('2023-07-04', '07:17', 100006223, 1, 'ASUdSgevAaADwlPcEwmdobbYrtIDlpLsLRYvfGXIaxCRukcnRWNCyruNTrPHcwGGktVmCfMqkGlDviMC'),('2023-07-04', '07:17', 100006222, 1, 'pDyoinDnkOKYnfpnzbnJCZWoJWlFiedMptmozuGGFDMnTEiixzsTpvOeUUklFAmWSZUGiGRJytnBRXdM'),('2023-07-04', '07:17', 100006207, 1, 'LvLqbHLpqgGbriXjbRghyKBXOtSTLQAVSrAtVCytdLbVCebOLvoGbVICHuVTvhGJEMLLpmNFHbzmUcPG'),('2023-07-04', '07:17', 100006200, 1, 'zoqEJUJUtnWmuYjBdVnFjjJafCSVbBjmmbmchVIiPjXAQzsxKCWmfntNRofpBNUEImXqArnIwLTcgfwD'),('2023-07-04', '07:17', 100006218, 1, 'rSAEqCCtvzqDDVqqGohrcTrGdwXgqXNTlABnlUNPAFNNwWxwjKdaoXTiuumdfIqoHDkUNHrtCOuYCQzE'),('2023-07-04', '07:17', 100006211, 1, 'dODuUaUFcTidIBveGrwocDiRkgUIxvbyYTsvSgZxYUigiUhsfsxRFUcuSQOkYomRUsrAfWYFZYkLNKrb'),('2023-07-04', '07:17', 100006224, 1, 'PKmjEPFpmMDMFxHOQaBFTRtcrMHXhSJZcjFpqIYprTcQGzEwHmSbCswBoMzERLkCoEVpLRzmqOouoAeI'),('2023-07-04', '07:17', 100006216, 1, 'DjEMHzoRfZphblpTFRIvpAVzzwdGMuYtMkJCviZzcxIJdyhsKATBxJRuifCVTfBTpxEsomueKKjAcJoi'),('2023-07-04', '07:17', 100006205, 1, 'PKHLGlhUhXmGnKyhkvIJliulmrYxvLYmAYagzGSGloxEOvdLjeFFdFtViVGKpoJdclDVvOXlWLbTlOHZ'),('2023-07-04', '07:17', 100006214, 1, 'oTiylmcHzFFIOJZlrOmjyTbubjksNajjaKTWsXydlAQWmlMtjdTfmFIKKzOAffELtjrdCRFNlLNqGgXw'),('2023-07-04', '07:17', 100006219, 1, 'pKFNEaLbPRkElTbxYClWNWWIiNkKeVMlvQUTAXfVTENLLRtOKVvTpZXRoYJxgWWERCruLHKecwVKNmmt'),('2023-07-04', '07:17', 100006201, 1, 'uWfbYIuGDfGsWmeNXNmKGjCmtPamoXnzvNaiEvwEiDBUnmoPCnCZNqBQKCxBsLDFAjSFatOWdKWNdCoQ'),('2023-07-04', '07:17', 100006217, 1, 'OnUQbQXUBxIcGNmvNINtuRXrGKPheBcRCzOyIfhEYUfiaQnFZUiJjUaGoMPTTjGQdYBxccPkkgKosnKm'),('2023-07-04', '07:17', 100006220, 1, 'OezLNOEigTGyXPyJFQbWAEOyfZynhuYtrikmmeGmaZfXRDwRKMorLslKXncQKnlClyvpnUZYIddVEvsw'),('2023-07-04', '07:17', 100006213, 1, 'ApNLCnQQKrORzmUDQJDTMKiAtebNXzqAYROQuZUmXSbVvbxkWCAvMUDHNkzbyfsvqGiqkAxxbljooOFD'),('2023-07-04', '07:17', 100006204, 1, 'ZNPXJNQfQVOABAenIWLqQRNXDGhPvEXgoUFLFwxIFmYnnxVcfvtWgCnfpuSESDlPPliaGdqVyXTBTZdN'),('2023-07-04', '07:17', 100006202, 1, 'gBaHqcBlwoSEQywQxkDgFwYzfQsCeDToyUBwzKZBUhNRmFONCpROKDOnFLIPWgFpigslfPzkYNsSAapZ'),('2023-07-04', '07:17', 100006212, 1, 'eDIGpMYxkOMobwiWFfDkUVEmxYfLbjILGRVpfKDcCZjXYYlLPwYKTyMYsDIOiBWoFLqBNSIFaiffshOS'),('2023-07-04', '07:17', 100006210, 1, 'asYattrZOBWkvuwESMAaRirLYmlSNMswfridhWtSBoidoqTakNzmYkaTEkAeEbsVwBhNxiIsZFbmNdWT'),('2023-07-04', '07:17', 100006208, 1, 'rIfYsDgwItogsIspcrMySJwDaVbOOwhfWYJDDGAgxUEWIHpkeTDFzDqbXqvBprZGXOnWnYchllotUoYT'),('2023-07-04', '07:17', 100006221, 1, 'mTOzUuipKLTlFLoFMSPeHIwqWNGNchtsYFbKJGXKfwEHenJXbHQrAbeLfFwcAzuQdjqvUTKIcCmqsWTD'),('2023-07-04', '07:17', 100006203, 1, 'UaJMOcmNAxBzsFUHaUFETADGTEAPVDJCTwIlUUTZcuwdhNQXUFZcQjYDFpemfkSjHlkVNqxMzeDgChQu'),('2023-07-04', '07:17', 100006215, 1, 'tAgUmvxXoGHVHbmQyWzvHbPwDaHbsHynzcAqbrCJjuHfUAlUdtRwlIrOyqXDoTmrckpzWrEepFmBBGNg'),('2023-07-04', '07:17', 100006209, 1, 'flEiDXNVQxgJYMWeDTLWyBGUBesIXHFOGXvWeMpGmEOokkauJpAHgKTuEPsYNazRDCMigIIEByZJkzCK'),('2023-07-04', '07:17', 100006206, 1, 'RedSEzDAuvwlkRPzfxXCYkJqVmrLZYODJPNrComlASSDcjgYNlAUqRYpuivdOdARkxifJbgSAKfxmsbY') on duplicate key update m_req_all=m_req_all+values(m_req_all);"] [err="runtime error: index out of range [0] with length 0"] [stack="goroutine 576252375 [running]:\ngithub.com/pingcap/tidb/server.(*clientConn).Run.func1(0x45116f0, 0xc015b3ced0, 0xc0059de780)\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/server/conn.go:1029 +0xf5\npanic(0x3d4b600, 0xc072c36630)\n\t/usr/local/go/src/runtime/panic.go:965 +0x1b9\ngithub.com/pingcap/tidb/executor.(*ExecStmt).Exec.func1(0xc03a79b860, 0xc051faf760, 0xc051faf740)\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/adapter.go:362 +0x4d4\npanic(0x3d4b600, 0xc072c36630)\n\t/usr/local/go/src/runtime/panic.go:965 +0x1b9\ngithub.com/pingcap/tidb/executor.getOldRow(0x45116f0, 0xc074ff6900, 0x455d378, 0xc00b711000, 0x4557610, 0xc00b711010, 0x7fd6440832c8, 0xc000cd4b40, 0x4545958, 0xc0568ca0f8, ...)\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/batch_checker.go:265 +0x9df\ngithub.com/pingcap/tidb/executor.(*InsertExec).updateDupRow(0xc09358f020, 0x45116f0, 0xc074ff6900, 0x0, 0x4557610, 0xc00b711010, 0xc07b933e00, 0x7, 0x7, 0x0, ...)\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/insert.go:193 +0xdd\ngithub.com/pingcap/tidb/executor.(*InsertExec).batchUpdateDupRows(0xc09358f020, 0x45116f0, 0xc074ff6900, 0xc048318500, 0x19, 0x19, 0x0, 0x0)\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/insert.go:270 +0x565\ngithub.com/pingcap/tidb/executor.(*InsertExec).exec(0xc09358f020, 0x45116f0, 0xc074ff6900, 0xc048318500, 0x19, 0x19, 0x0, 0x0)\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/insert.go:81 +0x77f\ngithub.com/pingcap/tidb/executor.insertRows(0x45116f0, 0xc074ff6900, 0x44e9740, 0xc09358f020, 0x3cbdd80, 0xc050a49960)\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/insert_common.go:278 +0x392\ngithub.com/pingcap/tidb/executor.(*InsertExec).Next(0xc09358f020, 0x45116f0, 0xc074ff67b0, 0xc0953b0a00, 0x0, 0xc051faf238)\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/insert.go:314 +0x92\ngithub.com/pingcap/tidb/executor.Next(0x45116f0, 0xc074ff67b0, 0x4516b20, 0xc09358f020, 0xc0953b0a00, 0x0, 0x0)\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/executor.go:286 +0x2de\ngithub.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor(0xc03a79b860, 0x45116f0, 0xc074ff67b0, 0x4516b20, 0xc09358f020, 0x0, 0x0, 0x0, 0x0)\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/adapter.go:635 +0x33c\ngithub.com/pingcap/tidb/executor.(*ExecStmt).handlePessimisticDML(0xc03a79b860, 0x45116f0, 0xc074ff67b0, 0x4516b20, 0xc09358f020, 0xc07164c433d67349, 0xf3a1f620e8ab)\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/adapter.go:654 +0x188\ngithub.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelay(0xc03a79b860, 0x45116f0, 0xc074ff67b0, 0x4516b20, 0xc09358f020, 0x644bb01, 0x4511601, 0x0, 0x0, 0x0, ...)\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/adapter.go:488 +0x155\ngithub.com/pingcap/tidb/executor.(*ExecStmt).Exec(0xc03a79b860, 0x45116f0, 0xc074ff67b0, 0x0, 0x0, 0x0, 0x0)\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/adapter.go:439 +0x72e\ngithub.com/pingcap/tidb/session.(*session).retry(0xc00b711000, 0x45116f0, 0xc030264e10, 0xa, 0x0, 0x0)\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/session/session.go:1051 +0x84c\ngithub.com/pingcap/tidb/session.(*session).doCommitWithRetry(0xc00b711000, 0x45116f0, 0xc030264e10, 0x0, 0x0)\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/session/session.go:833 +0x1377\ngithub.com/pingcap/tidb/session.(*session).CommitTxn(0xc00b711000, 0x45116f0, 0xc09a5b4540, 0x0, 0x0)\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/session/session.go:881 +0x145\ngithub.com/pingcap/tidb/session.autoCommitAfterStmt(0x45116f0, 0xc09a5b4540, 0xc00b711000, 0x0, 0x0, 0x4528720, 0xc03a79b860, 0x1, 0xc051fb09e0)\n\t/home"]

### 4. What is your TiDB version? (Required)
/ # ./tidb-server -V
Release Version: v5.4.0
Edition: Community
Git Commit Hash: 3679d5d28fb242fc9f047fbe57abb0243aaf9199
Git Branch: heads/refs/tags/v5.4.0
UTC Build Time: 2022-01-13 10:03:41
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: close #31735

Problem Summary:
In the insert statement, we only fill the public columns. So when calling getOldRow, we should not fill the generated column at the non-public state.

### What is changed and how it works?
Check column state when filling the generated column in getOldRow.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix a bug that insert statement may panic when adding a virtual generated column 
```

### CODE REVIEW

**Review by bb7133 — APPROVED** [review]
LGTM

File: ddl/db_test.go [hunk 1]
```diff
@@ -7711,3 +7711,52 @@ func (s *testDBSuite2) TestCreateTables(c *C) {
 	err = d.BatchCreateTableWithInfo(tk.Se, model.NewCIStr("test"), infos, ddl.OnExistError)
 	c.Check(terror.ErrorEqual(err, infoschema.ErrTableExists), IsTrue)
 }
+
+func (s *testSerialDBSuite) TestAddGeneratedColumnAndInsert(c *C) {
```
Comment (xiongjiwei) [inline ddl/db_test.go:7715]: we should use the new test framework

**Review by xiongjiwei — APPROVED** [review]

**Pull request merged**